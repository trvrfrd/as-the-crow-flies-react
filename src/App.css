* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

iframe {
  border: none;
  flex: 1;
  max-width: 1220px;
  width: 100%;
}

body {
  background-image: url('ross-parmly-25230-unsplash_2000_optim.jpg');
  background-position: right;
  background-size: cover;
  font-family: Frutiger, "Frutiger Linotype", Univers, Calibri, "Gill Sans", "Gill Sans MT", "Myriad Pro", Myriad, "DejaVu Sans Condensed", "Liberation Sans", "Nimbus Sans L", Tahoma, Geneva, "Helvetica Neue", Helvetica, Arial, sans-serif;
  font-size: 16px;
}

#app {
  display: flex;
  flex-direction: column;
  height: 100vh;
  width: 100vw;
}

header {
  align-items: center;
  display: flex;
  flex-basis: 80px;
  flex-shrink: 1;
  justify-content: center;
  margin-bottom: -10px;
}

main {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
}

h1 {
  font-size: calc(1.5em + 1.5vw);
}

h1, label, .itinerary {
  color: white;
  text-shadow: black 1px 1px 3px;
}

.typeahead {
  margin-top: 10px;
}

.typeahead fieldset {
  border: none;
}

.typeahead input {
  height: 2.2em;
  opacity: 0.8;
  text-align: center;
  width: 100%;
}

.suggestions {
  background-color: white;
  font-size: 0.8em;
  opacity: 0.8;
  position: absolute;
  width: 100%;
  z-index: 10;
}

.suggestion {
  border-bottom: 1px dashed black;
  border-left: 1px solid black;
  border-right: 1px solid black;
  padding: 5px;
}

.suggestion:last-child {
  border-bottom: 1px solid black;
}

.highlight {
  background-color: #FFEB3B;
  font-weight: bold;
  text-decoration: underline;
}

.itinerary {
  align-items: center;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  justify-content: center;
  text-align: center;
}

.itinerary p {
  font-size: calc(1em + 1vw);
  margin: 5px;
}

@media(min-width: 600px) {
  header {
    flex-basis: 100px;
  }

  .typeaheads {
    display: flex;
    justify-content: center;
  }

  .typeahead {
    flex-grow: 1;
    margin: 10px;
    max-width: 600px;
  }

  .suggestions {
    max-width: 600px;
    width: calc(50% - 20px);
  }

  .itinerary p {
    margin: 15px;
  }
}
