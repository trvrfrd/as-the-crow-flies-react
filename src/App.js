import React, { Component, Fragment } from 'react';

import Typeahead from './Typeahead';
import { calculateDistance } from './utils';
import { hot } from 'react-hot-loader';

import './App.css';

class App extends Component {
  state = {
    from: null,
    to: null,
    airports: []
  }

  componentDidMount() {
    this.props.loadAirports()
      .then(json => this.setState({ airports: json.airports }));
  }

  handleSelect = (endpoint, airport) => {
    this.setState({ [endpoint]: airport });
  }

  formatAirport = a => `(${a.iata}) ${a.name} - ${a.city}, ${a.state}`;

  render() {
    return (
      <Fragment>

        <header>
          <h1>As the Crow Flies</h1>
        </header>

        <main>
          <div className="typeaheads">
            <Typeahead
              name="from"
              placeholder="IATA code, airport name, city, or state"
              source={this.state.airports}
              maxSuggestions={10}
              formatSuggestion={this.formatAirport}
              onSelect={this.handleSelect.bind(this, 'from')}
            />

            <Typeahead
              name="to"
              placeholder="IATA code, airport name, city, or state"
              source={this.state.airports}
              maxSuggestions={10}
              formatSuggestion={this.formatAirport}
              onSelect={this.handleSelect.bind(this, 'to')}
            />
          </div>

          {
            this.state.from && this.state.to
            ? <Itinerary origin={this.state.from} destination={this.state.to} />
            : null
          }
        </main>

      </Fragment>
    );
  }
}

App.defaultProps = {
  loadAirports: () => import('./airports.json')
}

const Itinerary = ({ origin, destination}) => {
  const distance = calculateDistance(origin, destination).toFixed(0);
  return (
    <div className="itinerary">
      <p>
        {`${origin.iata} âœˆ ${destination.iata}: ${distance} nautical miles`}
      </p>
      <iframe
        src={mapsEmbedURL(origin, destination)}
      />
    </div>
  )
}

function mapsEmbedURL(origin, destination) {
  return 'https://www.google.com/maps/embed/v1/directions'
    + '?key=AIzaSyBz_m7FJ6G821bn-cmcLlmLzBaVc-xQquM'
    + `&origin=${encodeLatLon(origin)}`
    + `&destination=${encodeLatLon(destination)}`
    + '&mode=flying'
}

function encodeLatLon(place) {
  return encodeURIComponent(`${place.lat},${place.lon}`);
}

export default hot(module)(App);
