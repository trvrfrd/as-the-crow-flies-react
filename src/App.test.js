import React from 'react';
import { shallow, mount } from 'enzyme';
import App from './App';

const airportData = {
  airports: [
    {
      city: "New York",
      iata: "JFK",
      lat: 40.63980103,
      lon: -73.77890015,
      name: "John F Kennedy International Airport",
      state: "New York",
    },
    {
      city: "Los Angeles",
      iata: "LAX",
      lat: 33.94250107,
      lon: -118.4079971,
      name: "Los Angeles International Airport",
      state: "California",
    }
  ]
};

const loadAirports = () => Promise.resolve(airportData);

describe('App', () => {
  test('basic rendering snapshot', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  test('shows itinerary when both selections made', async () => {
    const wrapper = await mount(<App loadAirports={loadAirports} />);

    wrapper.find('#from').simulate('change', { target: { value: 'jfk' } });
    wrapper.find('.suggestion').first().simulate('click');
    wrapper.find('#to').simulate('change', { target: { value: 'lax' } });
    wrapper.find('.suggestion').first().simulate('click');

    expect(wrapper.find('.itinerary')).toHaveLength(1);
  })
});
