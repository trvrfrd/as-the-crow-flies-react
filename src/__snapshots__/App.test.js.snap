// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App basic rendering snapshot 1`] = `
Array [
  <header>
    <h1>
      As the Crow Flies
    </h1>
  </header>,
  <main>
    <div
      className="typeaheads"
    >
      <div
        className="typeahead"
      >
        <fieldset>
          <label
            htmlFor="from"
          >
            from
            : 
          </label>
          <br />
          <input
            id="from"
            name="from"
            onChange={[Function]}
            placeholder="IATA code, airport name, city, or state"
            value=""
          />
        </fieldset>
        <div
          className="suggestions"
        />
      </div>
      <div
        className="typeahead"
      >
        <fieldset>
          <label
            htmlFor="to"
          >
            to
            : 
          </label>
          <br />
          <input
            id="to"
            name="to"
            onChange={[Function]}
            placeholder="IATA code, airport name, city, or state"
            value=""
          />
        </fieldset>
        <div
          className="suggestions"
        />
      </div>
    </div>
  </main>,
]
`;
